File: classified/packages/agentserver/src/__tests__/integration/agent-export-import-e2e.test.ts
import { describe, it, expect, beforeAll, afterAll } from 'bun:test';
import { AgentServer } from '../../server';
import { AgentExportService } from '../../services/AgentExportService';
import { AgentImportService } from '../../services/AgentImportService';
import type { IAgentRuntime, UUID, Character } from '@elizaos/core';
import { AgentRuntime } from '@elizaos/core';
import { createDatabaseAdapter } from '@elizaos/plugin-sql';
import AdmZip from 'adm-zip';
import path from 'path';
import fs from 'fs';

describe('Agent Export/Import E2E Test', () => {
  let agentServer: AgentServer;
  let testDbPath: string;
  let sourceAgent: IAgentRuntime;
  let targetAgentId: UUID;

  beforeAll(async () => {
    // Use a test database
    testDbPath = path.join(__dirname, `test-db-export-import-${Date.now()}`);
    process.env.PGLITE_DATA_DIR = testDbPath;

    // Create and initialize agent server
    agentServer = new AgentServer();
    // Use PGLite URL format for testing
    const dbUrl = `pglite://${testDbPath}`;
    await agentServer.initialize(dbUrl);

    // Create a test agent
    const character: Character = {
      id: 'test-char-1' as UUID,
      name: 'Test Export Agent',
      bio: ['An agent for testing export/import'],
      topics: ['testing', 'export', 'import'],
      clients: [],
      plugins: [],
      settings: {
        model: 'gpt-4',
        secrets: {},
      },
      modelProvider: 'openai',
    };

    const db = createDatabaseAdapter({ dataDir: testDbPath }, 'test-agent-export-1' as UUID);

    await db.init();

    sourceAgent = new AgentRuntime({
      agentId: 'test-agent-export-1' as UUID,
      character,
      adapter: db,
      token: process.env.OPENAI_API_KEY || 'test-token',
      serverUrl: 'http://localhost:7777',
    } as any);

    // Register the agent
    await agentServer.registerAgent(sourceAgent);

    // Create some test data
    await db.createMemory({
      id: 'test-memory-1' as UUID,
      agentId: sourceAgent.agentId,
      content: { text: 'This is a test memory for export' },
      roomId: 'test-room-1' as UUID,
      entityId: 'test-entity-1' as UUID,
      unique: true,
      metadata: { type: 'test' },
    });

    targetAgentId = 'test-agent-import-1' as UUID;
  });

  afterAll(async () => {
    // Clean up
    if (agentServer.server) {
      await new Promise<void>((resolve) => {
        agentServer.server.close(() => resolve());
      });
    }

    // Clean up test database
    if (fs.existsSync(testDbPath)) {
      fs.rmSync(testDbPath, { recursive: true, force: true });
    }
  });

  it('should export and import agent data successfully', async () => {
    // Step 1: Export the agent
    const exportService = new AgentExportService(sourceAgent.agentId, sourceAgent, agentServer);

    const zipStream = await exportService.exportToZip();

    // Collect the stream into a buffer
    const chunks: Buffer[] = [];
    await new Promise<void>((resolve, reject) => {
      zipStream.on('data', (chunk) => chunks.push(chunk));
      zipStream.on('end', () => resolve());
      zipStream.on('error', reject);
    });

    const zipBuffer = Buffer.concat(chunks);
    await exportService.cleanup();

    // Verify the ZIP structure
    const zip = new AdmZip(zipBuffer);
    const entries = zip.getEntries();
    const entryNames = entries.map((e) => e.entryName);

    expect(entryNames).toContain('manifest.json');
    expect(entryNames).toContain('database/agent.json');
    expect(entryNames).toContain('database/memories.json');

    // Step 2: Import to a new agent
    const importService = new AgentImportService(targetAgentId, agentServer);

    const result = await importService.importFromZip(zipBuffer, {
      overwrite: true,
      validateManifest: true,
    });

    await importService.cleanup();

    // Verify import result
    expect(result.success).toBe(true);
    expect(result.agentId).toBe(targetAgentId);
    expect(result.agentName).toBe('Test Export Agent');
    expect(result.tablesImported).toContain('agent');
    expect(result.tablesImported).toContain('memories');

    // Step 3: Verify the imported data
    const targetDb = createDatabaseAdapter({ dataDir: testDbPath }, targetAgentId);

    await targetDb.init();

    // Check if memory was imported
    const memories = await targetDb.getMemories({
      agentId: targetAgentId,
      unique: true,
      count: 10,
    });

    expect(memories.length).toBe(1);
    expect(memories[0].content.text).toBe('This is a test memory for export');
  });

  it('should handle export of non-existent agent', async () => {
    const nonExistentId = 'non-existent-agent' as UUID;
    const exportService = new AgentExportService(nonExistentId, {} as IAgentRuntime, agentServer);

    await expect(exportService.exportToZip()).rejects.toThrow('Agent non-existent-agent not found');
    await exportService.cleanup();
  });

  it('should validate import manifest', async () => {
    const invalidZip = new AdmZip();
    invalidZip.addFile('test.txt', Buffer.from('invalid content'));

    const importService = new AgentImportService(targetAgentId, agentServer);

    await expect(importService.importFromZip(invalidZip.toBuffer())).rejects.toThrow(
      'Invalid export: manifest.json not found'
    );

    await importService.cleanup();
  });
});

File: cpp_merge/common/agent-export-import-e2e.test.cpp
#include "/home/runner/work/elizaos-cpp/elizaos-cpp/classified/packages/agentserver/src/__tests__/integration/agent-export-import-e2e.test.h"

void Main(void)
{
    describe(std:("Agent Export/Import E2E Test"), [=]() mutable
    {
        shared<std::shared_ptr<AgentServer>> agentServer;
        shared<string> testDbPath;
        shared<std::shared_ptr<IAgentRuntime>> sourceAgent;
        shared<std::shared_ptr<UUID>> targetAgentId;
        beforeAll([=]() mutable
        {
            testDbPath = path->join(__dirname, std:("test-db-export-import-") + Date->now() + string_empty);
            process->env->PGLITE_DATA_DIR = testDbPath;
            agentServer = std::make_shared<AgentServer>();
            auto dbUrl = std:("pglite://") + testDbPath + string_empty;
            std::async([=]() { agentServer->initialize(dbUrl); });
            auto character = object{
                object::pair{std:("id"), as<std::shared_ptr<UUID>>(std:("test-char-1"))}, 
                object::pair{std:("name"), std:("Test Export Agent")}, 
                object::pair{std:("bio"), array<string>{ std:("An agent for testing export/import") }}, 
                object::pair{std:("topics"), array<string>{ std:("testing"), std:("export"), std:("import") }}, 
                object::pair{std:("clients"), array<any>()}, 
                object::pair{std:("plugins"), array<any>()}, 
                object::pair{std:("settings"), object{
                    object::pair{std:("model"), std:("gpt-4")}, 
                    object::pair{std:("secrets"), object{}}
                }}, 
                object::pair{std:("modelProvider"), std:("openai")}
            };
            auto db = createDatabaseAdapter(object{
                object::pair{std:("dataDir"), testDbPath}
            }, as<std::shared_ptr<UUID>>(std:("test-agent-export-1")));
            std::async([=]() { db->init(); });
            sourceAgent = std::make_shared<AgentRuntime>(as<any>(object{
                object::pair{std:("agentId"), as<std::shared_ptr<UUID>>(std:("test-agent-export-1"))}, 
                object::pair{std:("character"), std:("character")}, 
                object::pair{std:("adapter"), db}, 
                object::pair{std:("token"), OR((process->env->OPENAI_API_KEY), (std:("test-token")))}, 
                object::pair{std:("serverUrl"), std:("http://localhost:7777")}
            }));
            std::async([=]() { agentServer->registerAgent(sourceAgent); });
            std::async([=]() { db->createMemory(object{
                object::pair{std:("id"), as<std::shared_ptr<UUID>>(std:("test-memory-1"))}, 
                object::pair{std:("agentId"), sourceAgent->agentId}, 
                object::pair{std:("content"), object{
                    object::pair{std:("text"), std:("This is a test memory for export")}
                }}, 
                object::pair{std:("roomId"), as<std::shared_ptr<UUID>>(std:("test-room-1"))}, 
                object::pair{std:("entityId"), as<std::shared_ptr<UUID>>(std:("test-entity-1"))}, 
                object::pair{std:("unique"), true}, 
                object::pair{std:("metadata"), object{
                    object::pair{std:("type"), std:("test")}
                }}
            }); });
            targetAgentId = as<std::shared_ptr<UUID>>(std:("test-agent-import-1"));
        }
        );
        afterAll([=]() mutable
        {
            if (agentServer->server) {
                std::async([=]() { std::make_shared<Promise<void>>([=](auto resolve) mutable
                {
                    agentServer->server->close([=]() mutable
                    {
                        return resolve();
                    }
                    );
                }
                ); });
            }
            if (fs->existsSync(testDbPath)) {
                fs->rmSync(testDbPath, object{
                    object::pair{std:("recursive"), true}, 
                    object::pair{std:("force"), true}
                });
            }
        }
        );
        it(std:("should and import agent data successfully"), [=]() mutable
        {
            auto exportService = std::make_shared<AgentExportService>(sourceAgent->agentId, sourceAgent, agentServer);
            shared zipStream = std::async([=]() { exportService->exportToZip(); });
            shared chunks = array<std::shared_ptr<Buffer>>();
            std::async([=]() { std::make_shared<Promise<void>>([=](auto resolve, auto reject) mutable
            {
                zipStream->on(std:("data"), [=](auto chunk) mutable
                {
                    return chunks->push(chunk);
                }
                );
                zipStream->on(std:("end"), [=]() mutable
                {
                    return resolve();
                }
                );
                zipStream->on(std:("error"), reject);
            }
            ); });
            auto zipBuffer = Buffer::concat(chunks);
            std::async([=]() { exportService->cleanup(); });
            auto zip = std::make_shared<AdmZip>(zipBuffer);
            auto entries = zip->getEntries();
            auto entryNames = entries->map([=](auto e) mutable
            {
                return e["entryName"];
            }
            );
            expect(entryNames)->toContain(std:("manifest.json"));
            expect(entryNames)->toContain(std:("database/agent.json"));
            expect(entryNames)->toContain(std:("database/memories.json"));
            auto importService = std::make_shared<AgentImportService>(targetAgentId, agentServer);
            auto result = std::async([=]() { importService->importFromZip(zipBuffer, object{
                object::pair{std:("overwrite"), true}, 
                object::pair{std:("validateManifest"), true}
            }); });
            std::async([=]() { importService->cleanup(); });
            expect(result->success)->toBe(true);
            expect(result->agentId)->toBe(targetAgentId);
            expect(result->agentName)->toBe(std:("Test Export Agent"));
            expect(result->tablesImported)->toContain(std:("agent"));
            expect(result->tablesImported)->toContain(std:("memories"));
            auto targetDb = createDatabaseAdapter(object{
                object::pair{std:("dataDir"), testDbPath}
            }, targetAgentId);
            std::async([=]() { targetDb->init(); });
            auto memories = std::async([=]() { targetDb->getMemories(object{
                object::pair{std:("agentId"), targetAgentId}, 
                object::pair{std:("unique"), true}, 
                object::pair{std:("count"), 10}
            }); });
            expect(memories->length)->toBe(1);
            expect(const_(memories)[0]->content->text)->toBe(std:("This is a test memory for export"));
        }
        );
        it(std:("should handle of non-existent agent"), [=]() mutable
        {
            auto nonExistentId = as<std::shared_ptr<UUID>>(std:("non-existent-agent"));
            auto exportService = std::make_shared<AgentExportService>(nonExistentId, as<std::shared_ptr<IAgentRuntime>>(object{}), agentServer);
            std::async([=]() { expect(exportService->exportToZip())->rejects->toThrow(std:("Agent non-existent-agent not found")); });
            std::async([=]() { exportService->cleanup(); });
        }
        );
        it(std:("should validate import manifest"), [=]() mutable
        {
            auto invalidZip = std::make_shared<AdmZip>();
            invalidZip->addFile(std:("test.txt"), Buffer::from(std:("invalid content")));
            auto importService = std::make_shared<AgentImportService>(targetAgentId, agentServer);
            std::async([=]() { expect(importService->importFromZip(invalidZip->toBuffer()))->rejects->toThrow(std:("Invalid export: manifest.json not found")); });
            std::async([=]() { importService->cleanup(); });
        }
        );
    }
    );
}

MAIN
